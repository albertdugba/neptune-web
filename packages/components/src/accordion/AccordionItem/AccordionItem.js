import React, { cloneElement } from 'react';
import Types from 'prop-types';
import classNames from 'classnames';
import { isString } from '@transferwise/neptune-validation';

import Chevron from '../../chevron';

const AccordionItem = ({ id, title, content, onClick, open, icon }) => {
  const iconEl = icon ? cloneElement(icon, { size: 24 }) : null;

  return (
    <div
      id={id}
      className={classNames('tw-accordion-item decision p-a-0', {
        closed: !open,
      })}
    >
      <button
        type="button"
        aria-expanded={open}
        onClick={onClick}
        className={classNames('tw-accordion-item btn btn-link p-x-0 text-no-decoration p-t-3', {
          'p-b-3': !open,
          'p-b-2': open,
        })}
      >
        <div className="media">
          {icon && <div className="media-left hidden-xs hidden-sm p-r-2">{iconEl}</div>}
          <div className="media-body text-xs-left title">
            {isString(title) ? <span className="h5">{title}</span> : title}
          </div>
          <div className="media-right d-flex align-items-center">
            <Chevron orientation={open ? Chevron.Orientation.TOP : Chevron.Orientation.BOTTOM} />
          </div>
        </div>
      </button>
      <div className="accordion-content media ">
        <div
          className={classNames('media-body m-r-5 p-b-3', { 'm-l-5': icon })}
          aria-hidden={!open}
        >
          {content}
        </div>
      </div>
    </div>
  );
};

AccordionItem.propTypes = {
  content: Types.node.isRequired,
  icon: Types.node,
  id: Types.string,
  open: Types.bool.isRequired,
  onClick: Types.func.isRequired,
  title: Types.node.isRequired,
};

AccordionItem.defaultProps = {
  icon: null,
  id: null,
};

export default AccordionItem;
