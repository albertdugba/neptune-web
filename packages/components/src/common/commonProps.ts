export type CommonProps = {
  /**
   * Space-separated list of the case-sensitive CSS classes,
   * most would be applied to outermost element inside the component
   *
   * @see https://github.com/transferwise/neptune-web/blob/main/rfc/0001-always-pass-classname.md
   */
  className?: string;
};

export type AriaLabelProp = {
  'aria-label'?: string;
};

export type LinkProps = {
  href: string;
  target?: '_blank' | '_self' | '_parent' | '_top';
};
