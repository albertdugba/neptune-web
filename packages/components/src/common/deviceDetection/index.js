export { isIosDevice } from './deviceDetection';
export { isTouchDevice } from './isTouchDevice';
