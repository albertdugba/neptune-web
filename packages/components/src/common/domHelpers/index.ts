export function isServerSideRendering(): boolean {
  return typeof window === 'undefined';
}

export {
  addClickClassToDocumentOnIos,
  removeClickClassFromDocumentOnIos,
} from './documentIosClick';
