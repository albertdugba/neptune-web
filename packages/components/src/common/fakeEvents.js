export const fakeEvent = () => ({
  nativeEvent: { stopImmediatePropagation() {} },
  stopPropagation() {},
  preventDefault() {},
});

export const fakeKeyDownEventForKey = (keyCode) => ({
  ...fakeEvent(),
  keyCode,
});

export const fakeKeyDownEvent = (eventAttribute = {}) => ({
  ...fakeEvent(),
  ...eventAttribute,
});
