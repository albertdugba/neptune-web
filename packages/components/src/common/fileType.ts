export enum FileType {
  PDF = '.pdf,application/pdf',
  JPG = '.jpg,.jpeg,image/jpeg',
  PNG = '.png,image/png',
  CSV = '.csv,text/csv',
}
