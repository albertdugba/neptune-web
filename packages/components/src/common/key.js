export const Key = {
  ENTER: 'Enter',
  ESCAPE: 'Escape',
  SPACE: [' ', 'Spacebar'],
  TAB: 'Tab',
  PERIOD: '.',
  COMMA: ',',
};
