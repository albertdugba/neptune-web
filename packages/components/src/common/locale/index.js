export const DEFAULT_LANG = 'en';
export const DEFAULT_LOCALE = 'en-GB';

export const SUPPORTED_RTL = ['he-IL'];

export const SUPPORTED_LANGUAGES = [
  DEFAULT_LANG,
  'ja',
  'de',
  'es',
  'fr',
  'hu',
  'id',
  'it',
  'pl',
  'pt',
  'ro',
  'ru',
  'tr',
  'zh',
];

/**
 * Verifies and adjusts locale (replace `_` with `-`)
 * Returns null if locale is unrecognized by {Intl.Locale}
 *
 * @param locale (`es`, `es_ES`, `en-GB`, `en`, `ja`, `ja-JP` etc)
 * @returns {Intl.Locale.baseName|null}
 */
export function adjustLocale(locale) {
  if (!locale || locale.trim().length === 0) {
    return null;
  }
  try {
    const { baseName } = new Intl.Locale(locale.trim().replace('_', '-'));
    return baseName;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    return null;
  }
}

/**
 * Provides corresponding lang (iso2) for provided locale
 * if locale is invalid or language is unsupported returns null
 *
 * @param locale (`es`, `es-ES`, `en-GB`, `en`, `ja`, `ja-JP` etc)
 * @returns {Intl.Locale.language|null}
 */
export function getLangFromLocale(locale) {
  const adjustedLocale = adjustLocale(locale);
  if (adjustedLocale === null) {
    return null;
  }
  try {
    const { language } = new Intl.Locale(adjustedLocale);
    if (SUPPORTED_LANGUAGES.includes(language)) {
      return language;
    }
    return null;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    return null;
  }
}

/**
 * Provides corresponding country code (iso2) for locales code with explicit region value (`es-ES`, `en-GB`, `ja-JP` etc.)
 * if the value is invalid or missing region it returns null
 *
 * @param locale
 * @returns {string|null}
 */
export const getCountryFromLocale = (locale) => {
  const adjustedLocale = adjustLocale(locale);
  if (adjustedLocale === null) {
    return null;
  }
  try {
    const { region } = new Intl.Locale(adjustedLocale);
    return region ?? null;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error(error);
    return null;
  }
};
