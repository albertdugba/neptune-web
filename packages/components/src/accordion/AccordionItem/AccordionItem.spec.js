import React from 'react';
import { Freeze } from '@transferwise/icons';

import { render, userEvent } from '../../test-utils';
import AccordionItem from './AccordionItem';

import { useDirection } from '../../common/hooks';

jest.mock('../../common/hooks');

describe('AccordionItem', () => {
  const props = {
    title: 'This is title number one',
    content: 'Lauri Ipsum has been the industry standard dummy text ever since the 1500s.',
    open: false,
    onClick: jest.fn(),
  };

  beforeAll(() => {
    useDirection.mockImplementation(() => ({ direction: 'ltr', isRTL: false }));
  });

  describe('open / close', () => {
    it('renders an item closed', () => {
      expect(render(<AccordionItem {...props} />).container).toMatchSnapshot();
    });

    it('renders an item open', () => {
      expect(render(<AccordionItem {...props} open />).container).toMatchSnapshot();
    });
  });

  describe('icons', () => {
    it('renders the icon if passed', () => {
      const { container } = render(<AccordionItem {...props} icon={<Freeze />} />);

      expect(container.querySelector('.tw-icon-freeze')).toBeInTheDocument();
    });

    it('will always render icons at 24px', () => {
      const { container } = render(<AccordionItem {...props} icon={<Freeze size={16} />} />);

      expect(container.querySelector('.tw-icon-freeze svg')).toHaveAttribute('width', '24');
    });
  });

  describe('onClick', () => {
    it('calls onClick when item is clicked', () => {
      const onClick = jest.fn();

      const { getByRole } = render(<AccordionItem {...props} onClick={onClick} />);

      userEvent.click(getByRole('button'));

      expect(onClick).toHaveBeenCalledTimes(1);
    });
  });
  describe('Right-To-Left', () => {
    it('applies correct css classes when isRTL is true', () => {
      useDirection.mockImplementation(() => ({ direction: 'rtl', isRTL: true }));
      const { container } = render(<AccordionItem {...props} />);
      expect(container.querySelector('.media-body')).toHaveClass('text-xs-right');
    });
  });
});
