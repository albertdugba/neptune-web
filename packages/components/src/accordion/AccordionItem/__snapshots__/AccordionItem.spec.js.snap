// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItem open / close renders an item closed 1`] = `
<div>
  <div
    class="tw-accordion-item p-a-0 tw-accordion-item--light closed"
  >
    <button
      aria-expanded="false"
      class="tw-accordion-item btn btn-link text-no-decoration decision"
      type="button"
    >
      <div
        class="media"
      >
        <div
          class="media-body title text-xs-left"
        >
          <span
            class="h5 tw-accordion-item__title"
          >
            This is title number one
          </span>
        </div>
        <div
          class="d-flex align-items-center media-right"
        >
          <span
            aria-hidden="true"
            class="tw-icon tw-icon-chevron-up tw-chevron chevron-color bottom"
            role="presentation"
          >
            <svg
              fill="currentColor"
              focusable="false"
              height="16"
              width="16"
            >
              <path
                clip-rule="evenodd"
                d="M8 4l-6.6 6.653L2.537 11.8 8 6.293l5.463 5.507 1.137-1.147L8 4z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </div>
    </button>
    <div
      class="accordion-content media"
    >
      <div
        aria-hidden="true"
        class="media-body p-t-2 p-b-3 p-l-2"
      >
        Lauri Ipsum has been the industry standard dummy text ever since the 1500s.
      </div>
    </div>
  </div>
</div>
`;

exports[`AccordionItem open / close renders an item open 1`] = `
<div>
  <div
    class="tw-accordion-item p-a-0 tw-accordion-item--light"
  >
    <button
      aria-expanded="true"
      class="tw-accordion-item btn btn-link text-no-decoration decision"
      type="button"
    >
      <div
        class="media"
      >
        <div
          class="media-body title text-xs-left"
        >
          <span
            class="h5 tw-accordion-item__title"
          >
            This is title number one
          </span>
        </div>
        <div
          class="d-flex align-items-center media-right"
        >
          <span
            aria-hidden="true"
            class="tw-icon tw-icon-chevron-up tw-chevron chevron-color top"
            role="presentation"
          >
            <svg
              fill="currentColor"
              focusable="false"
              height="16"
              width="16"
            >
              <path
                clip-rule="evenodd"
                d="M8 4l-6.6 6.653L2.537 11.8 8 6.293l5.463 5.507 1.137-1.147L8 4z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </div>
      </div>
    </button>
    <div
      class="accordion-content media"
    >
      <div
        aria-hidden="false"
        class="media-body p-t-2 p-b-3 p-l-2"
      >
        Lauri Ipsum has been the industry standard dummy text ever since the 1500s.
      </div>
    </div>
  </div>
</div>
`;
